Mathlib/NumberTheory/Padics/MahlerBasis.lean

```diff
@@ -95,3 +103,132 @@ The uniform norm of the `k`-th Mahler basis function is 1, for every `k`.
+    ← sub_eq_iff_eq_add.mpr (sum_range_succ' (fun x ↦ R.choose x • Δ_[1]^[n + x] f 0) R), add_zero,
+    Nat.choose_zero_right, one_smul]
+  have : ∑ k ∈ Finset.range (R + 1), R.choose k • Δ_[1]^[n + k] f 0 = Δ_[1]^[n] f R := by
+    simpa only [← Function.iterate_add_apply, add_comm, nsmul_one, add_zero] using
+      (shift_eq_sum_fwdDiff_iter 1 (Δ_[1]^[n] f) R 0).symm
+  simp only [this, fwdDiff_iter_eq_sum_shift (1 : M) f n, mul_comm, nsmul_one, mul_smul, add_comm,
+    add_zero, smul_sub, sum_sub_distrib]
+
+end fwdDiff
+
+namespace PadicInt
+
+section norm_fwdDiff
+
+variable {p : ℕ} [hp : Fact p.Prime] {E : Type*}
+  [NormedAddCommGroup E] [NormedSpace ℚ_[p] E] [IsUltrametricDist E]
+
+/--
+Second step in Bojanic's proof of Mahler's theorem (equation (11) of [bojanic74]): show that values
+`Δ_[1]^[n + p ^ t] f 0` for large enough `n` are bounded by the max of `(‖f‖ / p ^ s)` and `1 / p`
+times a sup over values for smaller `n`.
+
+We use `nnnorms` on the RHS since `Finset.sup` requires an order with a bottom element.
+-/
+private lemma bojanic_mahler_step2 {f : C(ℤ_[p], E)} {s t : ℕ}
+    (hst : ∀ x y : ℤ_[p], ‖x - y‖ ≤ p ^ (-t : ℤ) → ‖f x - f y‖ ≤ ‖f‖ / p ^ s) (n : ℕ) :
+    ‖Δ_[1]^[n + p ^ t] f 0‖ ≤ max ↑((Finset.range (p ^ t - 1)).sup
+      fun j ↦ ‖Δ_[1]^[n + (j + 1)] f 0‖₊ / p) (‖f‖ / p ^ s) := by
+  -- Use previous lemma to rewrite in a convenient form.
+  rw [bojanic_mahler_step1 _ _ (one_le_pow₀ hp.out.one_le)]
+  -- Now use ultrametric property and bound each term separately.
+  refine (norm_add_le_max _ _).trans (max_le_max ?_ ?_)
+  · -- Bounding the sum over `range (p ^ t - 1)`: every term involves a value `Δ_[1]^[·] f 0` and
+    -- a binomial coefficient which is divisible by `p`
+    rw [norm_neg, ← coe_nnnorm, coe_le_coe]
+    refine nnnorm_sum_le_of_forall_le (fun i hi ↦ Finset.le_sup_of_le hi ?_)
+    rw [mem_range] at hi
+    rw [← Nat.cast_smul_eq_nsmul ℚ_[p], nnnorm_smul, div_eq_inv_mul]
+    refine mul_le_mul_of_nonneg_right ?_ (by simp only [zero_le])
+    -- remains to show norm of binomial coeff is `≤ p⁻¹`
+    have : 0 < (p ^ t).choose (i + 1) := Nat.choose_pos (by omega)
+    rw [← zpow_neg_one, ← coe_le_coe, coe_nnnorm, Padic.norm_eq_pow_val (mod_cast this.ne'),
+      coe_zpow, NNReal.coe_natCast, (zpow_right_strictMono₀ (mod_cast hp.out.one_lt)).le_iff_le,
+      neg_le_neg_iff, Padic.valuation_natCast, Nat.one_le_cast]
+    exact one_le_padicValNat_of_dvd this <| hp.out.dvd_choose_pow (by omega) (by omega)
+  · -- Bounding the sum over `range (n + 1)`: every term is small by the choice of `t`
+    refine norm_sum_le_of_forall_le_of_nonempty nonempty_range_succ (fun i _ ↦ ?_)
+    rw [← Int.cast_smul_eq_zsmul ℚ_[p], norm_smul]
+    refine le_trans ?_ (hst (i + ↑(p ^ t)) i ?_)
+    · apply mul_le_of_le_one_left (norm_nonneg _)
+      simpa only [← coe_intCast] using norm_le_one _
+    · rw [Nat.cast_pow, add_sub_cancel_left, norm_pow, norm_p, inv_pow, zpow_neg, zpow_natCast]
```

Comment:
The code that's here works fine and has already been submitted for merging by a maintainer (failing to merge only because of a coincidental rename of a lemma). 

Can you explain why you believe all of these changes are so essential?

Reply:
The proof is now `calc`-ified. I don't personally think it makes it any clearer, but it's not worth arguing over.